src/lib/scalar.ts

```diff
@@ -0,0 +1,352 @@
+    constantValue ??= toConstantScalar(bits);
+    if (constantValue !== undefined) {
+      this.constantValue = ScalarConst.fromBigint(constantValue);
+    }
+  }
+
+  /**
+   * Create a constant {@link Scalar} from a bigint, number, string or Scalar.
+   *
+   * If the input is too large, it is reduced modulo the scalar field size.
+   */
+  static from(x: Scalar | Uint8Array | bigint | number | string) {
+    if (x instanceof Scalar) return x;
+    if (x instanceof Uint8Array) x = ScalarConst.toBigint(x);
+    let scalar = Fq(x);
+    let bits = toBits(scalar);
+    return new Scalar(bits, scalar);
+  }
+
+  /**
+   * Convert this {@link Scalar} into a constant if it isn't already.
+   *
+   * If the scalar is a variable, this only works inside `asProver` or `witness` blocks.
+   *
+   * See {@link FieldVar} for an explanation of constants vs. variables.
+   */
+  toConstant(): ConstantScalar {
+    if (this.constantValue !== undefined) return this as ConstantScalar;
+    let [, ...bits] = this.value;
+    let constBits = bits.map((b) => FieldVar.constant(Snarky.field.readVar(b)));
+    return new Scalar([0, ...constBits]) as ConstantScalar;
+  }
+
+  /**
+   * @deprecated use {@link Scalar.from}
+   */
+  static fromBigInt(x: bigint) {
+    return Scalar.from(x);
+  }
+
+  /**
+   * Convert this {@link Scalar} into a bigint
+   */
+  toBigInt() {
+    return this.#assertConstant('toBigInt');
+  }
+
+  // TODO: fix this API. we should represent "shifted status" internally and use
+  // and use shifted Group.scale only if the scalar bits representation is shifted
+  /**
+   * Creates a data structure from an array of serialized {@link Bool}.
+   *
+   * **Warning**: The bits are interpreted as the bits of 2s + 1 + 2^255, where s is the Scalar.
+   */
+  static fromBits(bits: Bool[]) {
+    return Scalar.fromFields(bits.map((b) => b.toField()));
+  }
+
+  /**
+   * Returns a random {@link Scalar}.
+   * Randomness can not be proven inside a circuit!
+   */
+  static random() {
+    return Scalar.from(Fq.random());
+  }
+
+  // operations on constant scalars
+
+  // TODO: this is a static method so that it works on ml scalars as well
+  static #assertConstantStatic(x: Scalar, name: string): Fq {
+    if (x.constantValue === undefined)
+      throw Error(
+        `Scalar.${name}() is not available in provable code.
+That means it can't be called in a @method or similar environment, and there's no alternative implemented to achieve that.`
+      );
+    return ScalarConst.toBigint(x.constantValue);
+  }
+
+  #assertConstant(name: string) {
+    return Scalar.#assertConstantStatic(this, name);
+  }
+
+  /**
+   * Negate a scalar field element.
+   *
+   * **Warning**: This method is not available for provable code.
+   */
+  neg() {
+    let x = this.#assertConstant('neg');
+    let z = Fq.negate(x);
+    return Scalar.from(z);
+  }
+
+  /**
+   * Add scalar field elements.
+   *
+   * **Warning**: This method is not available for provable code.
```

Comment:
Interesting - is this because Scalars are over Vesta and we do our provable operations over Pallas?